File: iOS-marker-grow/ViewController.m
//
//  ViewController.m
//  iOS-marker-grow
//
//  Created by 翁乐 on 29/11/2016.
//  Copyright © 2016 Autonavi. All rights reserved.
//

#import "ViewController.h"
#import "growAnnotationView.h"

@interface ViewController ()<MAMapViewDelegate>
{
    MAMapView *_mapView;
}

@end

@implementation ViewController

- (MAAnnotationView *)mapView:(MAMapView *)mapView viewForAnnotation:(id<MAAnnotation>)annotation
{
    if ([annotation isKindOfClass:[MAPointAnnotation class]])
    {
        static NSString *customReuseIdentifier = @"customReuseIdentifier";
        
        growAnnotationView *annotationView = (growAnnotationView *)[mapView dequeueReusableAnnotationViewWithIdentifier:customReuseIdentifier];
        
        if (annotationView == nil)
        {
            annotationView = [[growAnnotationView alloc] initWithAnnotation:annotation reuseIdentifier:@"growView"];
            
            annotationView.image = [UIImage imageNamed:@"CYAN"];
            annotationView.centerOffset = CGPointMake(0, -(annotationView.image.size.height / 2.0));
            
        }
        
        return annotationView;
    }
    
    return nil;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    _mapView = [[MAMapView alloc] initWithFrame:self.view.bounds];
    _mapView.delegate = self;
    _mapView.allowsAnnotationViewSorting = NO;
    
    [self.view addSubview:_mapView];
    
    UIButton *button = [[UIButton alloc] initWithFrame:CGRectMake(10, 100, 120, 32)];
    button.backgroundColor = [UIColor whiteColor];
    button.layer.borderWidth = 1.0;
    button.titleLabel.font = [UIFont systemFontOfSize:12];
    button.layer.borderColor = [UIColor redColor].CGColor;
    [button setTitleColor:[UIColor redColor] forState:UIControlStateNormal];
    [button setTitle:@"addAnnotation" forState:UIControlStateNormal];
    
    [button addTarget:self action:@selector(actionAddAnnotation) forControlEvents:UIControlEventTouchUpInside];
    
    [self.view addSubview:button];

}

- (void)actionAddAnnotation
{
    CGPoint randomPoint = CGPointZero;
    
    randomPoint.x = arc4random() % (int)(CGRectGetWidth(self.view.bounds) - 100);
    randomPoint.y = arc4random() % (int)(CGRectGetHeight(self.view.bounds) - 200);

    CLLocationCoordinate2D randomCoordinate = [_mapView convertPoint:randomPoint toCoordinateFromView:self.view];
    
    MAPointAnnotation *annotation = [[MAPointAnnotation alloc] init];
    annotation.coordinate = randomCoordinate;
    
    [_mapView addAnnotation:annotation];
}

@end

File: iOS-marker-grow-swift/ViewController.swift
//
//  ViewController.swift
//  iOS-marker-grow-swift
//
//  Created by shaobin on 16/12/6.
//  Copyright © 2016年 Autonavi. All rights reserved.
//

import UIKit

class ViewController: UIViewController, MAMapViewDelegate {
    
    var mapView : MAMapView!

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        self.mapView = MAMapView(frame: self.view.bounds)
        self.mapView.delegate = self
        self.mapView.allowsAnnotationViewSorting = false
        
        self.view.addSubview(self.mapView)
        
        let button = UIButton(frame: CGRect(x: CGFloat(10), y: CGFloat(100), width: CGFloat(120), height: CGFloat(32)))
        button.backgroundColor = UIColor.white
        button.layer.borderWidth = 1.0
        button.titleLabel?.font = UIFont.systemFont(ofSize: CGFloat(12))
        button.layer.borderColor = UIColor.red.cgColor
        button.setTitleColor(UIColor.red, for: .normal)
        button.setTitle("addAnnotation", for: .normal)
        button.addTarget(self, action: #selector(self.actionAddAnnotation), for: .touchUpInside)
        self.view.addSubview(button)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    @objc func actionAddAnnotation() {
        var randomPoint = CGPoint.zero
        randomPoint.x = CGFloat(Int(arc4random()) % Int(self.view.bounds.width - 100))
        randomPoint.y = CGFloat(Int(arc4random()) % Int(self.view.bounds.height - 200))
        let randomCoordinate: CLLocationCoordinate2D = mapView.convert(randomPoint, toCoordinateFrom: self.view)
        let annotation = MAPointAnnotation()
        annotation.coordinate = randomCoordinate
        mapView.addAnnotation(annotation)
    }
    
    //-MARK: mapview delegate
    
    func mapView(_ mapView: MAMapView!, viewFor annotation: MAAnnotation!) -> MAAnnotationView! {
        if(annotation.isKind(of: MAPointAnnotation.self)) {
            let reuseIdentifier = "yourReuseIdentifier"
            var annotationView:growAnnotationView! = mapView.dequeueReusableAnnotationView(withIdentifier: reuseIdentifier) as! growAnnotationView!
            
            if(annotationView == nil) {
                annotationView = growAnnotationView.init(annotation: annotation, reuseIdentifier: "yourReuseIdentifier")
                
                annotationView.image = UIImage(named: "CYAN")
                annotationView.centerOffset = CGPoint(x: 0, y: -(annotationView.image.size.height / 2.0))
            }
            
            
            return annotationView
        }
        
        return nil
    }
}

